### PR TITLE
prov/sm2: Implement CMA protocol and HMEM

### PR DESCRIPTION
**Pull request by sunkuamzn**
This PR implements (again) the CMA protocol for arbitrary sized host-host message transfers. It also implements the IPC protocol for arbitrary sized device->host/device transfers and a IPC fallback protocol for host->device transfers.

### CODE REVIEW

File: prov/sm2/src/sm2.h [hunk 1]
```diff
@@ -304,6 +316,16 @@ static inline size_t sm2_pop_xfer_entry(struct sm2_ep *ep,
 	return FI_SUCCESS;
 }
 
+static inline bool sm2_proto_imm_send_comp(uint16_t proto)
+{
+	switch (proto) {
```
Comment (a-szegel) [inline prov/sm2/src/sm2.h:331]: I think this could be written better as:
```
if (inject) {
   return true
}
return false
```

My gut instinct is telling me switch statements have more overhead than if, but I would need to verify with godbolt (might be wrong).

File: prov/sm2/src/sm2_ep.c [hunk 2]
```diff
@@ -228,6 +261,7 @@ static void cleanup_shm_resources(struct sm2_ep *ep)
 		} else {
 			/* TODO Tell other side that we haven't processed their
 			 * message, just returned xfer_entry */
+			xfer_entry->hdr.proto_flags &= ~SM2_GENERATE_COMPLETION;
```
Comment (a-szegel) [inline prov/sm2/src/sm2_ep.c:322]: I think this should be in a different commit, and that we should add an error flag to generate error completion.

File: prov/sm2/src/sm2_progress.c [hunk 3]
```diff
@@ -351,8 +438,9 @@ void sm2_progress_recv(struct sm2_ep *ep)
 					0, atomic_entry->atomic_data.data,
 					xfer_entry->hdr.size);
 			}
-			if (xfer_entry->hdr.op_flags & FI_DELIVERY_COMPLETE &&
-			    !(xfer_entry->hdr.proto_flags & SM2_UNEXP)) {
+
+			if (xfer_entry->hdr.proto_flags &
```
Comment (a-szegel) [inline prov/sm2/src/sm2_progress.c:725]: I would really like to hide the return handling logic in a static inline function for the following reasons:
1. it looks horrible with the 80 char line length limit
2. I think it makes the sm2 progress loop go from readable to very unclean.

File: prov/sm2/src/sm2_msg.c [hunk 4]
```diff
@@ -38,8 +38,34 @@
...
+static inline int sm2_select_proto(void **desc, size_t iov_count,
+				   uint64_t op_flags, uint64_t total_len)
 {
+	struct ofi_mr *sm2_desc;
+	enum fi_hmem_iface iface;
+
+	if (desc && *desc) {
+		sm2_desc = (struct ofi_mr *) *desc;
+		iface = sm2_desc->iface;
+		if (iface == FI_HMEM_CUDA) {
```
Comment (a-szegel) [inline prov/sm2/src/sm2_msg.c:50]: I think we should add an else with a warn printout "Non CUDA devices not supported"

File: prov/sm2/src/sm2_progress.c [hunk 5]
```diff
@@ -293,10 +438,93 @@ static int sm2_progress_recv_msg(struct sm2_ep *ep,
...
+	uint64_t comp_flags;
 	int ret;
 
+	/* TODO - Switch on protocol before switching on op to avoid messy
+	 * checks like this */
+
+	/* We don't want to look for rx entries for xfer entries with
+	 * SM2_CMA_HOST_TO_DEV and SM2_CMA_HOST_TO_DEV_ACK flags*/
+	if (xfer_entry->hdr.proto == sm2_proto_cma) {
+		if (xfer_entry->hdr.proto_flags & SM2_CMA_HOST_TO_DEV) {
```
Comment (a-szegel) [inline prov/sm2/src/sm2_progress.c:453]: I would like it if lines 453 -> 522 lived in a static inline function like sm2_host_to_dev_progress (feel free to come up with better name).  This is already messy... no need to make the common path unreadable.

File: prov/sm2/src/sm2_progress.c [hunk 6]
```diff
@@ -192,8 +329,8 @@ static int sm2_start_common(struct sm2_ep *ep,
 			rx_entry->iov, rx_entry->count, &total_len, ep);
 		break;
 	case sm2_proto_cma:
-		err = sm2_progress_cma(xfer_entry, rx_entry->iov,
-				       rx_entry->count, &total_len, ep, 0);
+		err = sm2_progress_cma(ep, xfer_entry, rx_entry, &total_len, 0,
+				       &ipc_host_to_dev);
```
Comment (a-szegel) [inline prov/sm2/src/sm2_progress.c:333]: I would like to add one if statement here that handles all the device to host stuff for this function, and returns early... so we do not have to mess up the rest of the common path with "!if device to host"

**Review by a-szegel — CHANGES_REQUESTED** [review]
Looks better... some minor changes please:

File: prov/sm2/src/sm2.h (hunk 1 above)
Comment (a-szegel) [reply prov/sm2/src/sm2.h:331]: if statement: https://godbolt.org/z/j8aW6novo

File: prov/sm2/src/sm2.h [hunk 7]
```diff
@@ -96,6 +96,17 @@ enum {
 
 #define SM2_RETURN (1 << 0)
 
+/* The SM2_UNEXP flag is used for unexpected receives. On the receiver, it
+indicates that the xfer_entry that the receiver is processing is actually in the
+xfer_ctx_pool and not in the receiver's freestack.
+
+For protocols that require delivery complete semantics (CMA and IPC), the
+receiver pops an xfer_entry from its own freestack, sets SM2_UNEXP flag together
+with SM2_RETURN and SM2_GENERATE_COMPLETION and sends it to the sender. The
```
[the remark this answers is not in the document]
Comment (sunkuamzn) [reply prov/sm2/src/sm2.h:107]: done

File: prov/sm2/src/sm2_ep.c (hunk 2 above)
Comment (sunkuamzn) [reply prov/sm2/src/sm2_ep.c:322]: done

File: prov/sm2/src/sm2_progress.c (hunk 3 above)
Comment (sunkuamzn) [reply prov/sm2/src/sm2_progress.c]: done

File: prov/sm2/src/sm2_msg.c (hunk 4 above)
Comment (sunkuamzn) [reply prov/sm2/src/sm2_msg.c:50]: Host memory can also have memory descriptors. So we would need to add another check for FI_HMEM_SYSTEM and that doesn't seem worth it to me.

File: prov/sm2/src/sm2_progress.c (hunk 5 above)
Comment (sunkuamzn) [reply prov/sm2/src/sm2_progress.c]: Done

File: prov/sm2/src/sm2_progress.c (hunk 6 above)
Comment (sunkuamzn) [reply prov/sm2/src/sm2_progress.c:333]: Done

**Comment by sunkuamzn** [comment]
The first commit is in https://github.com/ofiwg/libfabric/pull/9165 and will be rebased away when that PR is merged

**Review by a-szegel — APPROVED** [review]

**Comment by sunkuamzn** [comment]
bot:aws:retest

File: prov/sm2/src/sm2_ep.c [hunk 8]
```diff
@@ -228,6 +228,7 @@ static void cleanup_shm_resources(struct sm2_ep *ep)
...
```
Comment (aingerson) [inline prov/sm2/src/sm2_ep.c:322]: Maybe these patches are out of order and this will become clear in the next commits but in the previous commit you added the flag but never use it and in this one you just unset it?

File: prov/sm2/src/sm2_atomic.c [hunk 9]
```diff
@@ -183,15 +182,24 @@ static inline ssize_t sm2_generic_atomic(
...
-		ret = sm2_complete_tx(ep, context, op, op_flags);
-		if (ret)
-			FI_WARN(&sm2_prov, FI_LOG_EP_CTRL,
-				"unable to process tx completion\n");
+	/* Generate immediate send completion for ofi_op_atomic unless
+	 * FI_DELIVERY_COMPLETE are set */
+	switch (op) {
+	case ofi_op_atomic:
+		if (!(op_flags & FI_DELIVERY_COMPLETE)) {
+			ret = sm2_complete_tx(ep, context, op, op_flags);
```
Comment (aingerson) [inline prov/sm2/src/sm2_atomic.c:195]: I think we need a different check here. We can't complete yet for read requests

File: prov/sm2/src/sm2_progress.c [hunk 10]
```diff
@@ -329,6 +329,21 @@ static int sm2_progress_atomic(struct sm2_ep *ep,
...
 out:
+
+	/* Send completion for ofi_op_atomic is generated immediately on the
+	 * sender (unless FI_DELIVERY_COMPLETE flag is set). Other ops require
+	 * delivery complete semantics, so set the SM2_GENERATE_COMPLETION flag
+	 */
+	switch (xfer_entry->hdr.op) {
+	case ofi_op_atomic:
+		if (xfer_entry->hdr.op_flags & FI_DELIVERY_COMPLETE)
+			xfer_entry->hdr.proto_flags |= SM2_GENERATE_COMPLETION;
```
Comment (aingerson) [inline prov/sm2/src/sm2_progress.c:716]: I'm a little confused about the usage of these two flags. The commit message says that one flag is basically replacing the other but you're using them together? I think the use cases for the op_flags vs proto_flags needs to be outlined a little better to make it clearer what each is used for

File: prov/sm2/src/sm2.h [hunk 11]
```diff
@@ -102,6 +103,11 @@ enum {
 /* The SM2_UNEXP flag is used for unexpected receives. On the receiver, it
 indicates that the xfer_entry that the receiver is processing is actually in the
 xfer_ctx_pool and not in the receiver's freestack.
+
+For protocols that require delivery complete semantics (CMA and IPC), the
+receiver pops an xfer_entry from its own freestack, sets SM2_UNEXP flag together
+with SM2_RETURN and SM2_GENERATE_COMPLETION and sends it to the sender. The
```
Comment (aingerson) [inline prov/sm2/src/sm2.h:110]: This is a little confusing. Are you saying that's for the unexpected path? Setting UNEXP with GENERATE_COMPLETION seems contradictory

File: prov/sm2/src/sm2.h [hunk 12]
```diff
@@ -125,7 +131,7 @@ xfer_ctx_pool and not in the receiver's freestack.
  * 	proto - sm2 operation
  * 	proto_flags - Flags used by the sm2 protocol
  * 	sender_gid - id of msg sender
- * 	user_data - the message
+ * 	user_data - the message, for sm2_proto_inject
```
Comment (aingerson) [inline prov/sm2/src/sm2.h:139]: add description for cma

File: prov/sm2/src/sm2.h [hunk 13]
```diff
@@ -308,6 +319,16 @@ static inline size_t sm2_pop_xfer_entry(struct sm2_ep *ep,
 	return FI_SUCCESS;
 }
 
+static inline bool sm2_proto_imm_send_comp(uint16_t proto)
```
Comment (aingerson) [inline prov/sm2/src/sm2.h:332]: I haven't seen where this is used yet but it seems like a rather silly function and implementation for
`proto == sm2_proto_cma`
This also seems confusing given the earlier implementation of the generate completion flag/delivery complete flag. not all non cma will be able to generate an immediate send completion, right?
maybe it becomes more important in other patches or you're anticipating other protocols but for now I would drop this function and add a simpler local check

File: prov/sm2/src/sm2_progress.c [hunk 14]
```diff
@@ -42,6 +42,58 @@
...
+
+static int sm2_progress_cma(struct sm2_xfer_entry *xfer_entry,
+			    struct iovec *iov, size_t iov_count,
+			    size_t *total_len, struct sm2_ep *ep, int err)
+{
+	struct sm2_cma_data *cma_data =
+		(struct sm2_cma_data *) xfer_entry->user_data;
+	struct sm2_av *sm2_av =
+		container_of(ep->util_ep.av, struct sm2_av, util_av);
+	struct sm2_ep_allocation_entry *entries =
```
Comment (aingerson) [inline prov/sm2/src/sm2_progress.c]: use new `sm2_mmap_entries` function

File: prov/sm2/src/sm2_progress.c [hunk 15]
```diff
@@ -110,8 +166,35 @@ static int sm2_start_common(struct sm2_ep *ep,
...
+		 * generate a send completion */
+		ret = sm2_pop_xfer_entry(ep, &new_xfer_entry);
+		if (ret) {
+			FI_WARN(&sm2_prov, FI_LOG_EP_CTRL,
+				"Unable to send xfer_entry back to "
+				"sender. Sender has requested "
+				"FI_DELIVERY_COMPLETE but will not be "
+				"able to generate a send "
+				"completion!\n");
+			return ret;
```
Comment (aingerson) [inline prov/sm2/src/sm2_progress.c:391]: These seems like a dangerous way to handle this issue... just error out the sender because the app is sending too many things of its own? If you want to keep that fix for the later patch that fixes the unexpected messaging, that's ok, but maybe include that in your comment regarding the unexpected messaging handling future plans.

File: prov/sm2/src/sm2_progress.c [hunk 16]
```diff
@@ -120,6 +120,56 @@ static int sm2_progress_inject(struct sm2_xfer_entry *xfer_entry,
...
+
+	if (hmem_copy_ret < 0) {
+		FI_WARN(&sm2_prov, FI_LOG_EP_CTRL,
+			"IPC recv failed with code %d\n",
+			(int) (-hmem_copy_ret));
+		err = hmem_copy_ret;
+	} else if (hmem_copy_ret != xfer_entry->hdr.size) {
+		FI_WARN(&sm2_prov, FI_LOG_EP_CTRL, "IPC recv truncated\n");
+		err = -FI_ETRUNC;
+	} else
```
Comment (aingerson) [inline prov/sm2/src/sm2_progress.c]: brackets here

File: prov/sm2/src/sm2_progress.c [hunk 17]
```diff
@@ -120,6 +120,56 @@ static int sm2_progress_inject(struct sm2_xfer_entry *xfer_entry,
...
+		err = ret;
+
+	*total_len = hmem_copy_ret;
+
+	if (err)
+		return err;
+	else
+		return FI_SUCCESS;
```
Comment (aingerson) [inline prov/sm2/src/sm2_progress.c]: `return err`

File: prov/sm2/src/sm2_progress.c [hunk 18]
```diff
@@ -42,20 +42,136 @@
...
+
+	dest = (char *) (uintptr_t) mr_entry->info.mapped_addr +
+	       (uintptr_t) ipc_info->offset;
+
+	ret = ofi_copy_to_hmem(ipc_info->iface, ipc_info->device, dest,
+			       cma_data->iov[0].iov_base, xfer_entry->hdr.size);
+
+	ofi_mr_cache_delete(domain->ipc_cache, mr_entry);
+
+	return ret ? ret : FI_SUCCESS;
```
Comment (aingerson) [inline prov/sm2/src/sm2_progress.c:148]: `return ret`

File: prov/sm2/src/sm2_progress.c [hunk 19]
```diff
@@ -42,20 +42,136 @@
...
+
+static inline int sm2_cma_loop(struct sm2_ep *ep,
+			       struct sm2_xfer_entry *xfer_entry,
+			       struct fi_peer_rx_entry *rx_entry, size_t total,
 			       bool write)
 {
 	ssize_t ret;
+	struct sm2_av *sm2_av =
+		container_of(ep->util_ep.av, struct sm2_av, util_av);
+	struct sm2_ep_allocation_entry *entries =
```
Comment (aingerson) [inline prov/sm2/src/sm2_progress.c:159]: use `sm2_mmap_entries` call

File: prov/sm2/src/sm2_atomic.c [hunk 20]
```diff
@@ -109,9 +109,9 @@ sm2_do_atomic_inject(struct sm2_ep *ep, int64_t peer_gid, uint32_t op,
 		return ret;
 
 	xfer_entry->hdr.proto = sm2_proto_inject;
-	xfer_entry->hdr.proto_flags |= proto_flags;
 
 	sm2_generic_format(xfer_entry, ep->gid, op, 0, 0, op_flags, context);
+	xfer_entry->hdr.proto_flags |= proto_flags;
```
Comment (aingerson) [inline prov/sm2/src/sm2_atomic.c:114]: seems like this should just be fixed in the first patch

**Review by aingerson — COMMENTED** [review]
I didn't do a very in-depth review since sm2 development is being discontinued for the moment but enough to catch some potential issues. I think the generate completion vs unexp vs delivery complete flag usage is extremely confusing and is prone to a lot of issues. I'm ok with just adding comments to outline the future direction for if/when development continues because I think fixing it would be an invasive procedure and I know you're just wanting to get this in to preserve what you have worked on instead of losing it completely.

File: prov/sm2/src/sm2.h [hunk 21]
```diff
@@ -95,6 +95,9 @@ enum {
 	sm2_proto_max,
 };
 
+/* Protocol flags */
+#define SM2_ATOMIC_RMA_REQ (1 << 1)
```
[the remark this answers is not in the document]
Comment (sunkuamzn) [reply prov/sm2/src/sm2.h]: Done

File: prov/sm2/src/sm2.h [hunk 22]
```diff
@@ -99,6 +99,12 @@ enum {
 
 #define SM2_ATOMIC_RMA_REQ (1 << 1)
 
+/* The SM2_UNEXP flag is used for unexpected receives. On the receiver, it
+indicates that the xfer_entry that the receiver is processing is actually in the
+xfer_ctx_pool and not in the receiver's freestack.
+*/
```
[the remark this answers is not in the document]
Comment (sunkuamzn) [reply prov/sm2/src/sm2.h:114]: Not sure what you mean here. The sender is as agnostic as possible to whether the receive is unexpected. If the sender has requested delivery complete, then the sender can't generate completion until the application has posted the receive and the receiver has copied the data into the buffer. So the sender needs to be aware of the unexpected receive in case of delivery complete semantics.

The unexpected flow is different for different protocols, so I didn't add all of it in this commit. I added comments for the specific protocols in those commits. See e.g. https://github.com/ofiwg/libfabric/blob/8e8a706a4e7d978b4a3e4a3b704b2659a5475f63/prov/sm2/src/sm2.h#L107-L112

File: prov/sm2/src/sm2_progress.c [hunk 23]
```diff
@@ -355,7 +355,8 @@ void sm2_progress_recv(struct sm2_ep *ep)
 					0, atomic_entry->atomic_data.data,
 					xfer_entry->hdr.size);
 			}
-			if (xfer_entry->hdr.op_flags & FI_DELIVERY_COMPLETE) {
+			if (xfer_entry->hdr.op_flags & FI_DELIVERY_COMPLETE &&
+			    !(xfer_entry->hdr.proto_flags & SM2_UNEXP)) {
```
[the remark this answers is not in the document]
Comment (sunkuamzn) [reply prov/sm2/src/sm2_progress.c]: This was a bug in SM2. I fixed it in the next commit. This and previous commits are just adding/removing flags.

File: prov/sm2/src/sm2_ep.c (hunk 8 above)
Comment (sunkuamzn) [reply prov/sm2/src/sm2_ep.c:322]: Squashed this commit into the previous one

File: prov/sm2/src/sm2_atomic.c (hunk 9 above)
Comment (sunkuamzn) [reply prov/sm2/src/sm2_atomic.c:195]: Do you mean RMA read? RMA is not in SM2

File: prov/sm2/src/sm2.h (hunk 11 above)
Comment (sunkuamzn) [reply prov/sm2/src/sm2.h]: Updated comment

File: prov/sm2/src/sm2.h (hunk 12 above)
Comment (sunkuamzn) [reply prov/sm2/src/sm2.h]: Done

File: prov/sm2/src/sm2.h (hunk 13 above)
Comment (sunkuamzn) [reply prov/sm2/src/sm2.h:335]: It's also used for IPC. I think it's worth having because it's used for at least two protocols. https://github.com/ofiwg/libfabric/blob/e996e60f5d1cfb9253dcb6486ae0a84db14f2500/prov/sm2/src/sm2.h#L327-L336

Unexpected messages in other protocols can't generate completion if delivery complete is requested, yes. I can't think of any other exceptions.

File: prov/sm2/src/sm2_progress.c (hunk 14 above)
Comment (sunkuamzn) [reply prov/sm2/src/sm2_progress.c]: Done

File: prov/sm2/src/sm2_progress.c (hunk 15 above)
Comment (sunkuamzn) [reply prov/sm2/src/sm2_progress.c:394]: Makes sense. Add a TODO to add a retry.

File: prov/sm2/src/sm2_atomic.c (hunk 20 above)
Comment (sunkuamzn) [reply prov/sm2/src/sm2_atomic.c:114]: The first patch introduced a new proto flag but this patch fixed a bug in how atomics are using proto flags. I think they are different enough to be in separate commits. I can squash them if you want.

File: prov/sm2/src/sm2_progress.c (hunk 23 above)
[the remark this answers is not in the document]
Comment (sunkuamzn) [reply prov/sm2/src/sm2_progress.c]: The correct logic
https://github.com/ofiwg/libfabric/blob/8e8a706a4e7d978b4a3e4a3b704b2659a5475f63/prov/sm2/src/sm2_progress.c#L459-L473

File: prov/sm2/src/sm2.h (hunk 22 above)
[the remark this answers is not in the document]
Comment (aingerson) [reply prov/sm2/src/sm2.h:114]: Delivery complete just means the data is somewhere in the receiver (whether the target buffer or a bounce buffer). For the case of an unexpected message, this means we can copy the payload into an internal buffer so when the receiver actually posts the buffer, the data is actually coming from an internal location rather than the command. This is how shm does it right now for inline, inject, and sar. The sender is completely unaware whether the data is in the posted receive or not.

File: prov/sm2/src/sm2_atomic.c (hunk 9 above)
Comment (aingerson) [reply prov/sm2/src/sm2_atomic.c:195]: No, I meant an atomic with FI_ATOMIC_READ but I see this is only for the base atomics, and the atomic read is a fetch atomic so nevermind!
Looking at this again though, I think a single if statement here would work better instead of a switch.

File: prov/sm2/src/sm2.h (hunk 13 above)
Comment (aingerson) [reply prov/sm2/src/sm2.h:335]: I'm ok with merging it as is just to get it in but if we do resume development of sm2 I think we need to rework the completion flag usage. Sometimes the completion is triggered using the delivery complete flag and sometimes it's the generate completion flag and it's set later in the command and then we have this function that is confusing. I think we can make the decision up front with a single flag set and always check that flag.
For example:
```
#define SMR_COMPLETE_LATER

send_atomic()
{
if (!(atomic base && !(op_flags & FI_DELIVERY_COMPLETE))
    xfer_entry->proto_flags |= SMR_COMPLETE_LATER

if (!proto_flags & SMR_COMPLETE_LATER)
    complete_tx() 
}

progress_return()
{
//do return things

if (proto_flags & SMR_COMPLETE_LATER)
    complete_tx()
```

File: prov/sm2/src/sm2_progress.c [hunk 24]
```diff
@@ -122,6 +122,54 @@ static int sm2_progress_inject(struct sm2_xfer_entry *xfer_entry,
...
+	struct ofi_mr_entry *mr_entry;
+	struct ipc_info *ipc_info = (struct ipc_info *) xfer_entry->user_data;
+
+	domain = container_of(ep->util_ep.domain, struct sm2_domain,
+			      util_domain);
+
+	ret = ofi_ipc_cache_search(domain->ipc_cache,
+				   xfer_entry->hdr.sender_gid, ipc_info,
+				   &mr_entry);
+
```
Comment (aingerson) [inline prov/sm2/src/sm2_progress.c:277]: Extra space here if you feel like removing it. If you're not going to update the patch set, don't worry about it.

File: prov/sm2/src/sm2_progress.c [hunk 25]
```diff
@@ -138,6 +186,11 @@ static int sm2_start_common(struct sm2_ep *ep,
 			xfer_entry, (struct ofi_mr **) rx_entry->desc,
 			rx_entry->iov, rx_entry->count, &total_len, ep, 0);
 		break;
+	case sm2_proto_ipc:
+		err = sm2_progress_ipc(
+			xfer_entry, (struct ofi_mr **) rx_entry->desc,
```
Comment (aingerson) [inline prov/sm2/src/sm2_progress.c:326]: Format looks funky

**Review by aingerson — COMMENTED** [review]
Just did a quick pass so we can get this in. A few lingering things that are not big issues or just need a comment in there for a future TODO or something.

File: prov/sm2/src/sm2.h (hunk 22 above)
[the remark this answers is not in the document]
Comment (sunkuamzn) [reply prov/sm2/src/sm2.h:114]: > Delivery complete just means the data is somewhere in the receiver (whether the target buffer or a bounce buffer).

That was not my understanding of delivery complete. While working on these changes, both @a-szegel  and I thought delivery complete meant that the data has to be in the receiver's application buffer. So this set of commits uses an unnecessarily higher standard and has quite a bit of logic to make that happen..

Are you OK with me just adding a comment about it for now?

File: prov/sm2/src/sm2_atomic.c (hunk 9 above)
Comment (sunkuamzn) [reply prov/sm2/src/sm2_atomic.c:195]: I can turn it into an `if`

File: prov/sm2/src/sm2_progress.c (hunk 25 above)
Comment (sunkuamzn) [reply prov/sm2/src/sm2_progress.c:326]: We are all at the mercy of clang-format gods

File: prov/sm2/src/sm2.h (hunk 22 above)
[the remark this answers is not in the document]
Comment (aingerson) [reply prov/sm2/src/sm2.h:114]: Yeah that's fine to just have a comment for now. Thanks!

File: prov/sm2/src/sm2.h (hunk 13 above)
Comment (sunkuamzn) [reply prov/sm2/src/sm2.h:335]: Agree. The atomics work was done independently, so we didn't think about using a common flag..

File: prov/sm2/src/sm2_progress.c (hunk 25 above)
Comment (sunkuamzn) [reply prov/sm2/src/sm2_progress.c:326]: The formatting actually got fixed in a later commit

File: prov/sm2/src/sm2.h (hunk 22 above)
[the remark this answers is not in the document]
Comment (j-xiong) [reply prov/sm2/src/sm2.h:114]: @sunkuamzn You might have mixed delivery complete with match complete.

[the remark this answers is not in the document]
Comment (sunkuamzn) [reply prov/sm2/src/sm2.h:114]: Yes I think we both made that mistake https://ofiwg.github.io/libfabric/main/man/fi_cq.3.html

[the remark this answers is not in the document]
Comment (a-szegel) [reply prov/sm2/src/sm2.h:114]: But the receiver's libfabric had to touch the message... it can't be sitting the the receivers fifo queue.  The receiver needs to do some processing, and return the FQE.

**Comment by j-xiong** [comment]
@aingerson Is it ready to be included in 1.20?

**Review by aingerson — APPROVED** [review]
@j-xiong Good go!
